feat: display costs for registration

AlexD10S opened the pull request:
The `pop up` command uses this functionality for deploying a parachain. Since this is an expensive operation, it's important to provide users with clear cost visibility, displaying both the estimated cost of the call and their available balance to ensure they have sufficient funds.

This PR introduces a cost summary for the calls: 
- Reserve a parachain id 
- Register a parachain
Calculating the deposits  for each operation and estimating the fees for the call. 
It also fetches and displays the user's balance, providing a clear warning message if the available funds are insufficient.

**To calculate the deposit:**
- For reserve call: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/common/src/paras_registrar/mod.rs#L600
- For register call: https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/runtime/common/src/paras_registrar/mod.rs#L683

This improvement is based on feedback from @chungquantin  https://github.com/r0gue-io/pop-cli/pull/410#pullrequestreview-2631761590 during the review of the `pop up` feature.

**For testing locally**
Is a bit tricky without `pop-cli`, you can build it and replace the index.html in `pop cli`. 
Or for a quick test locally I replaced the line 71 in `SigningPortal.tsx`:
```const result = await fetchPayload(); ```
 with
```
const result = {
        chain_rpc: "ws://127.0.0.1:59830",
        call_data: Binary.fromHex("0x4605").asBytes(),
 }
```


**Some screenshots:**
<img width="421" alt="Captura de pantalla 2025-02-25 a las 21 30 43" src="https://github.com/user-attachments/assets/4b13647b-6d5c-437f-923a-d15602721fbc" />
<img width="434" alt="Captura de pantalla 2025-02-25 a las 21 30 28" src="https://github.com/user-attachments/assets/246de200-bd8d-40b6-a67b-fdc222cd27d5" />


[sc-3090]

AlexD10S commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -103,22 +116,45 @@ export const SigningPortal: React.FC = () => {
-          setChainProperties(chainSpec.properties)
-        }
+        // Extract properties & provide default values for missing ones
+        let updatedProperties: ChainProperties = {
```
I had to fix the way chain properties are set because, in some cases, `tokenSymbol` is missing, preventing the properties from updating correctly. Now, the update ensures that missing values are handled properly.

chungquantin commented on src/CostSummary/CostSummary.tsx:
```diff
@@ -0,0 +1,63 @@
+      </div>
+      {isOpen && (
+        <div className="mt-3 space-y-2">
+          <div className="flex justify-between items-center">
```
Suggest refactoring to a reusable component that accepts `(title, amount, decimals, symbol, className)`.

AlexD10S replied:
I agree, cleaner! https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/918490a6cdb732d6e961fa0d2ab4b7993cbdb911

chungquantin commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -81,13 +86,21 @@ export const SigningPortal: React.FC = () => {
         let callData = Binary.fromHex(hex);
         setCallData(callData);
         const tx = await api.txFromCallData(callData);
+        // console.log(tx.decodedCall.type);
```
Should be these removed?

AlexD10S replied:
https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/76d80128bbf245c7e6f1184f6ed2649430a13a1f

chungquantin commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -81,13 +86,21 @@ export const SigningPortal: React.FC = () => {
+        // console.log(tx.decodedCall.type);
+        // console.log(tx.decodedCall.value.type);
         setTx(tx);
         setIsContract(tx.decodedCall.type === "Contracts");
```
Suggest assigning a variable for `tx.decodedCall.type`

AlexD10S replied:
https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/76d80128bbf245c7e6f1184f6ed2649430a13a1f

chungquantin commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -320,6 +388,7 @@ export const SigningPortal: React.FC = () => {
           <div className="pb-3">
             <div className="font-semibold pb-2">Extrinsic Info:</div>
             {tx ? (
+              <>
```
```suggestion
              <React.Fragment>
```
Not a blocker. I think using `React.Fragment` makes the code easier to read, feel free to apply if you like it. I'm okay with both styles.

AlexD10S replied:
Nice. https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/76d80128bbf245c7e6f1184f6ed2649430a13a1f

chungquantin reviewed (commented):
Overall the code looks great. I'll test the functionalities later but after that I'm happy to approve.

peterwht commented on src/CostSummary/CostSummary.tsx:
```diff
@@ -0,0 +1,53 @@
+            <div className="flex justify-between items-center">
+              <h4 className="text-md font-semibold">Your Balance:</h4>
+              <div className={`px-3 py-1 rounded font-bold ${isInsufficientFunds ? "bg-red-100 text-red-700 border border-red-400" : "bg-green-100 text-green-700 border border-green-400"}`}>
+                {formatCurrency(accountBalance, chainProperties.tokenDecimals)} {chainProperties.tokenSymbol}
```
would format with commas. See what Tin did here: https://github.com/r0gue-io/pop-website/pull/70/commits/80aa15ef73dbac652719e66f9d15821be8797440#diff-aa2c37d6f1cee1c50ff72d7e8867368ceceabfbd38c2403994f8a5b652aa4f3eR102

AlexD10S replied:
Based on your comment above consistency is not better to use `formatCurrency`? as it is the one used in the dry-run too?

AlexD10S replied:
Understood now, fixed in https://github.com/r0gue-io/pop-wallet-signing-portal/commit/ee80256e57e8cfacb312a336dc8a34d58a5c02c7

peterwht commented on src/CostSummary/CostSummary.tsx:
```diff
@@ -0,0 +1,53 @@
+            {isInsufficientFunds && (
+              <div className="p-3 bg-red-100 border-l-4 border-red-500 text-red-700 rounded">
+                <p className="font-semibold">Insufficient Funds</p>
+                <p>You need at least <span className="font-bold">{formatCurrency(totalCost, chainProperties.tokenDecimals)} {chainProperties.tokenSymbol}</span> to complete this transaction.</p>
```
similarly

AlexD10S replied:
https://github.com/r0gue-io/pop-wallet-signing-portal/commit/ee80256e57e8cfacb312a336dc8a34d58a5c02c7

peterwht commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -103,22 +117,46 @@ export const SigningPortal: React.FC = () => {
 
+  useEffect(() => {
+    const getSelectedAccountFreeBalance = async () => {
+      // @ts-ignore
```
reason for this?

AlexD10S replied:
Same as you used in `await api?.apis.ContractsApi.call`( https://github.com/r0gue-io/pop-wallet-signing-portal/blob/main/src/SigningPortal/SigningPortal.tsx#L164) to ignore type-checking error:
```
'api.query.System.Account' is of type 'unknown' is of type 'unknown'.
```

chungquantin replied:
I also have this for all similar cases in pop-website.

peterwht commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -197,6 +235,38 @@ export const SigningPortal: React.FC = () => {
+
+  const calculateCosts = async (tx: UnsafeTransaction<any, string, string, any>, api: UnsafeApi<any>) => {
+    let decodedCall = tx?.decodedCall;
+    if (!selectedAccount || (tx.decodedCall.type !== "Registrar" || (tx.decodedCall.value.type !== "reserve" && tx.decodedCall.value.type !== "register"))) {
```
might make sense to convert you conditional to a variable or function. (this one `(tx.decodedCall.type !== "Registrar" || (tx.decodedCall.value.type !== "reserve" && tx.decodedCall.value.type !== "register")`

Used a few times and is quite long.

AlexD10S replied:
Nice, refactored: https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/b8254baa6f920cce505a86c6b3852cfbe51f1dda

peterwht commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -197,6 +235,38 @@ export const SigningPortal: React.FC = () => {
+      case "register":
+        let args = decodedCall.value.value;
+        let genesisHead = args.genesis_head;
+        // @ts-ignore
```
reason for this?

AlexD10S replied:
Same as mentioned above

peterwht commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -197,6 +235,38 @@ export const SigningPortal: React.FC = () => {
+    let fees = await tx.getEstimatedFees(selectedAccount.address);
+    setFeesEstimation(fees);
+
+
```
```suggestion
```

AlexD10S replied:
https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/b8254baa6f920cce505a86c6b3852cfbe51f1dda

peterwht commented on src/SigningPortal/SigningPortal.tsx:
```diff
@@ -82,12 +87,21 @@ export const SigningPortal: React.FC = () => {
+        let dispatachable = tx.decodedCall.value.type;
+        setIsContract(pallet === "Contracts");
+        setIsChainRegistrar(pallet === "Registrar" && (dispatachable === "reserve" || dispatachable === "register"));
+        
```
```suggestion
```

AlexD10S replied:
https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/b8254baa6f920cce505a86c6b3852cfbe51f1dda

peterwht reviewed (changes requested):
Looking good, nice add.

The following is non-blocking those as it's a small design nitpick. So, up to your discretion based on timelines. But, I think the summary box is a little inconsistent with the contracts summary box.

When a contract call will be successful:
![Screenshot 2025-02-25 at 11 54 28 AM](https://github.com/user-attachments/assets/537b8418-e780-4bc3-a1f1-886eb80c12b8)
![Screenshot 2025-02-25 at 11 54 35 AM](https://github.com/user-attachments/assets/2e8c607c-bf6f-4c96-97d9-2611b52a763b)




failing
![Screenshot 2025-02-25 at 11 53 35 AM](https://github.com/user-attachments/assets/c68435d4-f63c-44c8-9527-3066eb9d51ef)

![Screenshot 2025-02-25 at 11 53 44 AM](https://github.com/user-attachments/assets/89847f00-c685-488f-a373-f7f958497478)




Not to say that's the best way, but you could copy the success and fails bit from the contracts box (or make it a component to wrap around the internal elements).

AlexD10S commented:
> Looking good, nice add.
> 
> The following is non-blocking those as it's a small design nitpick. So, up to your discretion based on timelines. But, I think the summary box is a little inconsistent with the contracts summary box.

I agree it should be more consistent with the existing one. I've fixed it in this commit: https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/4046b6da26006318035602b2d23f8eaba06dc5bc. 
I have updated the screenshots in the description.

chungquantin reviewed (commented):
To be clear, I don't think the signing portal will ever be used on mobile but just note it here that the current UI is not optimized for the smaller resolutions.  Depends on you whether this is important. And I agree with Peter that the number should be formatted with commas

<img width="304" alt="Screenshot 2025-02-26 at 17 09 41" src="https://github.com/user-attachments/assets/c126c466-424d-4a40-a7ae-bfb66f1c23f7" />

Aside from that, I really like how the balance is updated real time when I updated it.

chungquantin reviewed (approved):
Aside from Peter's comments, I don't have any other concerns or feedback so happy to approve! Thank you for this PR.

AlexD10S commented:
> And I agree with Peter that the number should be formatted with commas

Updated `formatCurrency` to use `toLocaleString`: https://github.com/r0gue-io/pop-wallet-signing-portal/pull/7/commits/ee80256e57e8cfacb312a336dc8a34d58a5c02c7

Note that for me, it's not showing a comma. In Europe, it's the other way around, so I agree with you both using `.toLocaleString()` seems the best solution.

peterwht reviewed (approved):
Nice job! Thanks for updating
